lang/cs/Org.Apache.REEF.Network/Group/Codec/StreamingGroupCommunicationMessageCodec.cs

```diff
@@ -0,0 +1,120 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using Org.Apache.REEF.Network.Group.Driver.Impl;
+using Org.Apache.REEF.Network.StreamingCodec;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Wake.Remote;
+using ProtoBuf;
+using Microsoft.Hadoop.Avro;
+using Org.Apache.REEF.Network.StreamingCodec.Impl;
+
+namespace Org.Apache.REEF.Network.Group.Codec
+{
+    /// <summary>
+    /// Used to serialize GroupCommunicationMessages. This is the StreamingCodec for 
+    /// Group Communication Message. The base data codec in DataPlusCodec can either 
+    /// be streaming or non-streaming.
+    /// </summary>
+    public class StreamingGroupCommunicationMessageCodec : StreamingDataCodec<StreamingGroupCommunicationMessage>
+    {
+        /// <summary>
+        /// Create a new GroupCommunicationMessageCodec.
+        /// </summary>
+        [Inject]
+        public StreamingGroupCommunicationMessageCodec()
+        {
+        }
+
+        /// <summary>
+        /// Serialize the GroupCommunicationObject into a byte array.
+        /// </summary>
+        /// <param name="obj">The object to serialize.</param>
+        /// <returns>The serialized GroupCommunicationMessage in byte array form</returns>
+        public override byte[] Encode(StreamingGroupCommunicationMessage obj)
+        {
+            MemoryStream stream = new MemoryStream();
+            Encode(obj, new StreamWithPosition(stream, 0));
+            return stream.ToArray();
+        }
+
+        /// <summary>
+        /// Deserialize the byte array into a GroupCommunicationMessage using Protobuf.
+        /// </summary>
+        /// <param name="data">The byte array to deserialize</param>
+        /// <returns>The deserialized GroupCommunicationMessage object.</returns>
+        public override StreamingGroupCommunicationMessage Decode(byte[] data)
+        {
+            MemoryStream memStream = new MemoryStream(data);
+            return Decode(new StreamWithPosition(memStream, 0));
+        }
+
+        /// <summary>Encodes the given StreamingGroupCommunicatioMessage object and 
+        /// write it to the stream. First all the fields except the data and codec are 
+        /// encoded using Avro. Then data is written using the codec.</summary>
+        /// <param name="obj"> the object to be encoded</param>
+        /// <param name="stream"> stream we want the encoding written to</param>
+        public override void Encode(StreamingGroupCommunicationMessage obj, StreamWithPosition stream)
+        {
+            var avroSerializer = AvroSerializer.Create<StreamingGroupCommunicationMessage>();
+            avroSerializer.Serialize(stream.CodecStream, obj);
+
+            if (obj.Data.Codec is IStreamingDataCodecObj)
```

Comment:
You could move all this logic into `StreamingGroupCommunicationMessage`, which in itself gets an `Encode` method. That way, we can potentially avoid the `if(){...}else{...}` altogether when we know the type of codec we use at compile time.